builtin/logical/mysql/path_role_create.go

```diff
@@ -51,9 +52,15 @@ func (b *backend) pathRoleCreateRead(
 	}
 
 	// Generate our username and password. MySQL limits user to 16 characters
-	displayName := req.DisplayName
-	if len(displayName) > 10 {
-		displayName = displayName[:10]
+	displayName := name
+	ul, ok := data.GetOk("username_length")
+	if ok == true {
+		usernameLength = ul.(int)
+	} else {
+		usernameLength = 10
+	}
+	if len(displayName) > usernameLength {
```

Comment:
16 characters is still a hard limit on MySQL versions <5.7.8, which is why I'm not increasing the default: http://dev.mysql.com/doc/refman/5.7/en/user-names.html

But you're right -- there should really be two limits in play here: role.UsernameLength and role.DisplaynameLength